payload/reggae/rules/cmake.d

```diff
@@ -0,0 +1,320 @@
+    });
+
+    return selectedCMakeTargets.filter!((target) {
+        if (target["type"].str.isSupportedTarget) {
+            return true;
+        }
+        stderr.writeln("[WARNING] Skipping unsupported target '" ~ target["name"].str ~
+                       "' of type '" ~ target["type"].str ~ "'");
+        return false;
+    }).map!(t => t.toReggaeTarget(cmakeInfo)).array;
+}
+
+private imported!"std.json".JSONValue readJson(in string path) {
+    import std.file : readText;
+    import std.json : parseJSON;
+    string content = readText(path);
+    return parseJSON(content);
+}
+
+private string findJsonFilePath(in string apiReplyDir, in string name) {
+    import std.exception : enforce;
+    import std.array : array;
+    import std.path : baseName, extension;
+    import std.file : dirEntries, SpanMode;
+    import std.algorithm : map, filter, startsWith;
+
+    auto files = dirEntries(apiReplyDir, SpanMode.shallow)
+                 .map!(de => de.name)
+                 .filter!(f => f.baseName.startsWith(name) && f.extension == ".json")
+                 .array;
+
+    enforce(files.length == 1, "Could not find the '" ~ name ~ "' JSON file.");
+    return files[0];
+}
+
+private enum SourceLanguage : string {
+    C = "C",
+    CPP = "CXX",
+}
+
+private bool isSupportedLanguage(in string lang) {
+    import std.traits : EnumMembers;
+    import std.algorithm : canFind;
+    return [EnumMembers!SourceLanguage].canFind(lang);
+}
+
+private struct CompileOptions {
+    import reggae.types : Flags, IncludePaths;
+    Flags compilerFlags;
+    IncludePaths includes;
+    SourceLanguage language;
+}
+
+private CompileOptions getCompileOptions(in imported!"std.json".JSONValue targetJsonObj, ulong compileGroupIndex) {
+    import std.format : format;
+    import std.exception : enforce;
+    import std.array : join, array;
+    import std.algorithm : map;
+
+    auto compileGroups = targetJsonObj["compileGroups"].array;
+    enforce(compileGroups.length > compileGroupIndex, "Index out of bounds.");
+
+    auto compileGroup = compileGroups[compileGroupIndex];
+    const language = compileGroup["language"].str;
+    enforce(isSupportedLanguage(language), "'%s' is not supported.".format(language));
+
+    immutable notSupportedFeatures = ["precompileHeaders"];
+    foreach (feature; notSupportedFeatures) {
+        enforce(feature !in compileGroup, "The '" ~ feature ~ "' is not supported.");
+    }
+
+    static CompileOptions getCompileOptionsImpl(in imported!"std.json".JSONValue compileGroup, ulong groupIndex) {
+        import reggae.types : IncludePaths, Flags;
+
+        static CompileOptions[ulong] groupIndexToOptions;
+
+        if (auto opt = groupIndex in groupIndexToOptions) {
+            return *opt;
+        }
+
+        string defines;
+        if ("defines" in compileGroup) {
+            defines = compileGroup["defines"].array
+                            .map!(jobj => "-D" ~ jobj["define"].str)
```

Comment:
I wonder if it's better to have separate fields in `CompileOptions` for defines, etc. and let someone else add the actual flags later. For instance, I'm pretty sure that this isn't the correct flag for cl.exe.

Reply:
Given that this is only required for the `defines`, I would say it's fine to solve it here.